lib/active_model/serializer/adapter/attributes.rb

```diff
@@ -2,42 +2,58 @@ module ActiveModel
   class Serializer
     module Adapter
       class Attributes < Base
+        attr_accessor :_nested_associations
+
         def serializable_hash(options = nil)
           options ||= {}
+
+          self._nested_associations = ActiveModel::Serializer::Utils
+            .include_args_to_hash(options[:include])
+
           if serializer.respond_to?(:each)
-            result = serializer.map { |s| Attributes.new(s).serializable_hash(options) }
+            serializable_hash_for_collection(options)
           else
-            hash = {}
+            serializable_hash_for_single_resource(options)
+          end
+        end
+
+        def serializable_hash_for_collection(options = nil)
+          serializer.map { |s| Attributes.new(s).serializable_hash(options) }
+        end
+
+        def serializable_hash_for_single_resource(options = {})
+          resource = resource_object_for(serializer, options)
+          add_resource_relationships(resource, serializer)
+        end
+
+        # iterate through the associations on the serializer,
+        # adding them to the parent as needed (as singular or plural)
+        #
+        # nested_associations is a list of symbols that governs what
+        # associations on the passed in seralizer to include
+        def add_resource_relationships(parent, serializer)
+          included_associations = relevant_included_associations_for(serializer, _nested_associations)
 
-            core = cache_check(serializer) do
-              serializer.attributes(options)
+          included_associations.each do |association|
+            serializer = association.serializer
+            opts = association.options
+            key = association.key
+
+            # sanity check if the association has nesting data
+            has_nesting = _nested_associations[key].present?
+            if has_nesting
+              include_options_from_parent = { include: _nested_associations[key] }
+              opts = opts.merge(include_options_from_parent)
             end
 
-            serializer.associations.each do |association|
-              serializer = association.serializer
-              association_options = association.options
-
-              if serializer.respond_to?(:each)
-                array_serializer = serializer
-                hash[association.key] = array_serializer.map do |item|
-                  cache_check(item) do
-                    item.attributes(association_options)
-                  end
-                end
-              else
-                hash[association.key] =
-                  if serializer && serializer.object
-                    cache_check(serializer) do
-                      serializer.attributes(options)
-                    end
-                  elsif association_options[:virtual_value]
-                    association_options[:virtual_value]
-                  end
-              end
+            if serializer.respond_to?(:each)
```

Comment:
I think the problem comes from the fact that when you do a recursive call, you have to explicitly set the `include` option, otherwise it will default to "all relationships", and give you infinite loops.